### PR TITLE
Skip modifying User-Agent in ApplyUserAgentStage if the user has already passed custom User-Agent in the request

### PR DESCRIPTION
**Pull request by joviegas**
## Motivation and Context

PR #6513 moved the `ApplyUserAgentStage` to execute after `MergeCustomHeadersStage`. This caused user-supplied User-Agent headers to be overwritten by the SDK's default User-Agent.

**Previous behavior:** User-supplied User-Agent headers would overwrite the SDK's User-Agent because `ApplyUserAgentStage` ran before `MergeCustomHeadersStage`.

**After PR #6513:** The SDK's User-Agent now overwrites user-supplied User-Agent headers because `ApplyUserAgentStage` runs after `MergeCustomHeadersStage`.

This PR fixes the issue by skipping User-Agent modification in `ApplyUserAgentStage` when a custom User-Agent is already provided via `ADDITIONAL_HTTP_HEADERS`.

## Modifications

- Added check in `ApplyUserAgentStage` to detect if User-Agent header exists in `ADDITIONAL_HTTP_HEADERS` configuration
- Skip SDK User-Agent application when custom User-Agent is present
- Added unit tests to verify User-Agent preservation behavior
- Added functional tests covering various User-Agent configuration scenarios

## Testing

- Added unit tests in `ApplyUserAgentStageTest` covering: - Custom User-Agent preservation - Empty/null User-Agent handling - User-Agent presence in additional headers
- Added functional tests in `CustomUserAgentHeaderTest` covering: - Single and multiple custom User-Agent values - API name handling with custom User-Agent - Edge cases (empty, null, list values)
- Local run of `mvn install` succeeds
- All existing tests pass , tested this with Older version to make sure we are backward compatible before PR #6513

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)

## License
- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: .changes/next-release/bugfix-AWSSDKforJavav2-f9f830e.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
+    "type": "bugfix",
+    "category": "AWS SDK for Java v2",
+    "contributor": "",
+    "description": "Skip User-Agent header modification in ApplyUserAgentStage when custom User-Agent is already provided."
```
Comment (dagnir) [inline .changes/next-release/bugfix-AWSSDKforJavav2-f9f830e.json]: This is heavy on internal details. Can we simplify so we just say we don't overwrite the custom `User-Agent`?

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/useragent/CustomUserAgentHeaderTest.java [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+import software.amazon.awssdk.services.restjsonendpointproviders.RestJsonEndpointProvidersClientBuilder;
+
+/**
+ * Functional tests verifying custom User-Agent header preservation.
+ *
+ * <p>Tests ensure that User-Agent headers provided via
+ * {@link software.amazon.awssdk.core.client.config.ClientOverrideConfiguration.Builder#putHeader(String, String)} are preserved
+ * and not overwritten by SDK's default User-Agent generation logic.
+ */
+class CustomUserAgentHeaderTest {
```
Comment (dagnir) [inline test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/useragent/CustomUserAgentHeaderTest.java:48]: Can we add a test for when the header is added using the request override config?

Comment (joviegas) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/useragent/CustomUserAgentHeaderTest.java:48]: Good catch , currently the test case fails to handle backward compatibility

File: .changes/next-release/bugfix-AWSSDKforJavav2-f9f830e.json (hunk 1 above)
Comment (joviegas) [reply .changes/next-release/bugfix-AWSSDKforJavav2-f9f830e.json]: done

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/useragent/CustomUserAgentHeaderTest.java (hunk 2 above)
Comment (joviegas) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/useragent/CustomUserAgentHeaderTest.java:48]: Done , added test cases and fix for this

Comment (dagnir) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/useragent/CustomUserAgentHeaderTest.java:48]: Just to double check, request override User-Agent used to take precedence before the user agent stage fix right?

File: .changes/next-release/bugfix-AWSSDKforJavav2-f9f830e.json (hunk 1 above)
Comment (dagnir) [reply .changes/next-release/bugfix-AWSSDKforJavav2-f9f830e.json]: Looks the same, forgot to commit the new file perhaps? 😅

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/useragent/CustomUserAgentHeaderTest.java (hunk 2 above)
Comment (joviegas) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/useragent/CustomUserAgentHeaderTest.java:48]: Hi Dongie

Yeap request override User-Agent takes the precedence.

Same as Client Overide config when User set User-Agent putHeader on RequestOverride config the  User set User-Agent will be preserved and sent to the service , this behaviour is same as before we moved the ApplyUserAgent

```java
    void executeRequest_withRequestLevelCustomUserAgent_shouldPreserveAndNotOverwrite(String customUserAgent) {
        RestJsonEndpointProvidersClient client = defaultClientBuilder().build();

        assertThatThrownBy(() -> client.allTypes(r -> r
            .overrideConfiguration(o -> o.putHeader(USER_AGENT_HEADER, customUserAgent))))
            .hasMessageContaining(INTERCEPTOR_STOP_MESSAGE);

        String userAgent = getCapturedUserAgent();
        assertThat(userAgent)
            .isEqualTo(customUserAgent)
            .doesNotContain(SDK_USER_AGENT_PREFIX);
    }
```

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/ApplyUserAgentStage.java [hunk 3]
```diff
@@ -66,10 +67,38 @@ public ApplyUserAgentStage(HttpClientDependencies dependencies) {
     @Override
     public SdkHttpFullRequest.Builder execute(SdkHttpFullRequest.Builder request,
                                               RequestExecutionContext context) throws Exception {
+
+        if (hasUserAgentInAdditionalHeaders() || hasUserAgentInRequestHeaders(context)) {
```
Comment (dagnir) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/ApplyUserAgentStage.java]: super nit: `hasUserAgentInRequestHeaders` "inRequestConfig", otherwise it sounds like the header is in the marshalled `SdkHttpRequest`

File: .changes/next-release/bugfix-AWSSDKforJavav2-f9f830e.json (hunk 1 above)
Comment (joviegas) [reply .changes/next-release/bugfix-AWSSDKforJavav2-f9f830e.json]: my bad

File: test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/useragent/CustomUserAgentHeaderTest.java (hunk 2 above)
Comment (joviegas) [reply test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/useragent/CustomUserAgentHeaderTest.java:48]: Tested independently with 
Backward compatibilty test passed
Below is 1 version before breaking chnage
```xml
            <groupId>software.amazon.awssdk</groupId>
            <artifactId>dynamodb</artifactId>
            <version>2.36.1</version>
        </dependency>
```

Backward compatibilty test failed
```xml
            <groupId>software.amazon.awssdk</groupId>
            <artifactId>dynamodb</artifactId>
            <version>2.40.1</version>
        </dependency>
```

```java
package com.example;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.core.interceptor.Context;
import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
import software.amazon.awssdk.core.interceptor.ExecutionInterceptor;
import software.amazon.awssdk.http.SdkHttpRequest;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
import software.amazon.awssdk.services.dynamodb.DynamoDbClientBuilder;

/**
 * Functional tests verifying custom User-Agent header preservation for DynamoDB client.
 */
class DynamoDbUserAgentTest {

    private static final String USER_AGENT_HEADER = "User-Agent";
    private static final String SDK_USER_AGENT_PREFIX = "aws-sdk-java";
    private static final String TEST_API_NAME = "TestApiName";
    private static final String TEST_API_VERSION = "1.0";
    private static final String INTERCEPTOR_STOP_MESSAGE = "stop";

    private CapturingInterceptor interceptor;

    private static Stream<Arguments> customUserAgentValues() {
        return Stream.of(
            Arguments.of("CustomUserAgentHeaderValue"),
            Arguments.of("MyApplication/1.0.0"),
            Arguments.of("CustomClient/2.0 (Linux; x86_64)")
        );
    }

    private static Stream<Arguments> customUserAgentListValues() {
        return Stream.of(
            Arguments.of(Arrays.asList("Agent1")),
            Arguments.of(Arrays.asList("Agent1", "Agent2")),
            Arguments.of(Arrays.asList("CustomClient/1.0", "MyApp/2.0"))
        );
    }

    @BeforeEach
    void setUp() {
        interceptor = new CapturingInterceptor();
    }

    // ========== Default Behavior Tests ==========

    @Test
    void executeRequest_withoutCustomUserAgent_shouldAddSdkDefaultUserAgent() {
        DynamoDbClient client = defaultClientBuilder().build();
        executeRequestExpectingInterception(client);

        assertUserAgentContains(SDK_USER_AGENT_PREFIX);
    }

    // ========== Custom User-Agent Preservation Tests ==========

    @ParameterizedTest(name = "Custom User-Agent ''{0}'' should be preserved without SDK prefix")
    @MethodSource("customUserAgentValues")
    void executeRequest_withCustomUserAgent_shouldPreserveAndNotOverwrite(String customUserAgent) {
        DynamoDbClient client = clientWithCustomUserAgent(customUserAgent);
        executeRequestExpectingInterception(client);

        String userAgent = getCapturedUserAgent();
        assertThat(userAgent)
            .isEqualTo(customUserAgent)
            .doesNotContain(SDK_USER_AGENT_PREFIX);
    }

    @ParameterizedTest(name = "Custom User-Agent list {0} should be preserved")
    @MethodSource("customUserAgentListValues")
    void executeRequest_withCustomUserAgentList_shouldPreserveAllValues(List<String> customUserAgentList) {
        DynamoDbClient client = clientWithCustomUserAgentList(customUserAgentList);
        executeRequestExpectingInterception(client);

        List<String> userAgentList = getCapturedUserAgentList();
        assertThat(userAgentList).isEqualTo(customUserAgentList);
    }

    // ========== Request-Level User-Agent Tests ==========

    @ParameterizedTest(name = "Request-level User-Agent ''{0}'' should be preserved")
    @MethodSource("customUserAgentValues")
    void executeRequest_withRequestLevelCustomUserAgent_shouldPreserveAndNotOverwrite(String customUserAgent) {
        DynamoDbClient client = defaultClientBuilder().build();

        assertThatThrownBy(() -> client.listTables(r -> r
            .overrideConfiguration(o -> o.putHeader(USER_AGENT_HEADER, customUserAgent))))
            .hasMessageContaining(INTERCEPTOR_STOP_MESSAGE);

        String userAgent = getCapturedUserAgent();
        assertThat(userAgent)
            .isEqualTo(customUserAgent)
            .doesNotContain(SDK_USER_AGENT_PREFIX);
    }

    @ParameterizedTest(name = "Request-level User-Agent list {0} should be preserved")
    @MethodSource("customUserAgentListValues")
    void executeRequest_withRequestLevelCustomUserAgentList_shouldPreserveAllValues(List<String> customUserAgentList) {
        DynamoDbClient client = defaultClientBuilder().build();

        assertThatThrownBy(() -> client.listTables(r -> r
            .overrideConfiguration(o -> o.putHeader(USER_AGENT_HEADER, customUserAgentList))))
            .hasMessageContaining(INTERCEPTOR_STOP_MESSAGE);

        List<String> userAgentList = getCapturedUserAgentList();
        assertThat(userAgentList).isEqualTo(customUserAgentList);
    }

    @Test
    void executeRequest_withRequestLevelCustomUserAgentAndApiName_shouldNotAppendApiName() {
        String customUserAgent = "CustomUserAgentHeaderValue";
        DynamoDbClient client = defaultClientBuilder().build();

        assertThatThrownBy(() -> client.listTables(r -> r
            .overrideConfiguration(o -> o
                .addApiName(api -> api.name(TEST_API_NAME).version(TEST_API_VERSION))
                .putHeader(USER_AGENT_HEADER, customUserAgent))))
            .hasMessageContaining(INTERCEPTOR_STOP_MESSAGE);

        String userAgent = getCapturedUserAgent();
        assertThat(userAgent)
            .isEqualTo(customUserAgent)
            .doesNotContain(TEST_API_NAME);
    }

    @ParameterizedTest(name = "Request-level User-Agent list {0} with API name should not append API name")
    @MethodSource("customUserAgentListValues")
    void executeRequest_withRequestLevelCustomUserAgentListAndApiName_shouldNotAppendApiName(List<String> customUserAgentList) {
        DynamoDbClient client = defaultClientBuilder().build();

        assertThatThrownBy(() -> client.listTables(r -> r
            .overrideConfiguration(o -> o
                .addApiName(api -> api.name(TEST_API_NAME).version(TEST_API_VERSION))
                .putHeader(USER_AGENT_HEADER, customUserAgentList))))
            .hasMessageContaining(INTERCEPTOR_STOP_MESSAGE);

        List<String> userAgentList = getCapturedUserAgentList();
        assertThat(userAgentList).isEqualTo(customUserAgentList);
        assertThat(String.join(" ", userAgentList)).doesNotContain(TEST_API_NAME);
    }

    // ========== Header via Interceptors ==========

    @Test
    void executeRequest_withInterceptorAddingUserAgent_shouldAddSdkDefaultUserAgent() {
        DynamoDbClient client =
            defaultClientBuilder().overrideConfiguration(o -> o
                .addExecutionInterceptor(interceptor)
                .addExecutionInterceptor(new ExecutionInterceptor() {
                    @Override
                    public SdkHttpRequest modifyHttpRequest(Context.ModifyHttpRequest context,
                                                            ExecutionAttributes executionAttributes) {
                        return context.httpRequest().toBuilder()
                                      .putHeader(USER_AGENT_HEADER, "custom-agent")
                                      .build();
                    }
                })).build();

        executeRequestExpectingInterception(client);
        assertUserAgentContains(SDK_USER_AGENT_PREFIX);
    }

    // ========== API Name Handling Tests ==========

    @Test
    void executeRequest_withCustomUserAgentAndApiName_shouldNotAppendApiName() {
        String customUserAgent = "CustomUserAgentHeaderValue";
        DynamoDbClient client = clientWithCustomUserAgent(customUserAgent);
        executeRequestWithApiName(client);

        String userAgent = getCapturedUserAgent();
        assertThat(userAgent)
            .isEqualTo(customUserAgent)
            .doesNotContain(TEST_API_NAME);
    }

    @Test
    void executeRequest_withoutCustomUserAgentAndWithApiName_shouldAppendApiName() {
        DynamoDbClient client = defaultClientBuilder().build();
        executeRequestWithApiName(client);

        assertUserAgentContains(TEST_API_NAME + "/" + TEST_API_VERSION);
    }

    // ========== Edge Case Tests ==========

    @Test
    void buildClient_withNullListUserAgent_shouldThrowNullPointerException() {
        assertThatThrownBy(() -> clientWithCustomUserAgentList(null))
            .isInstanceOf(NullPointerException.class)
            .hasMessageContaining("values must not be null");
    }

    @Test
    void executeRequest_withEmptyListUserAgent_shouldResultInSdkUserAgentHeader() {
        DynamoDbClient client = clientWithCustomUserAgentList(Collections.emptyList());
        executeRequestExpectingInterception(client);

        List<String> userAgentList = getCapturedUserAgentList();
        assertThat(userAgentList).isNull();
    }

    @Test
    void executeRequest_withEmptyCustomUserAgent_shouldStoreSdkUserAgent() {
        DynamoDbClient client = clientWithCustomUserAgent("");
        executeRequestExpectingInterception(client);

        assertUserAgentContains("");
    }

    @Test
    void executeRequest_withNullStringUserAgent_shouldStoreAsSdkUserAgent() {
        DynamoDbClient client = clientWithCustomUserAgent(null);
        executeRequestExpectingInterception(client);

        List<String> userAgentList = getCapturedUserAgentList();
        assertThat(userAgentList)
            .hasSize(1)
            .allSatisfy(ua -> {
                assertThat(ua).isNull();
            });
    }

    @Test
    void executeRequest_withRequestLevelEmptyCustomUserAgent_shouldStoreEmptyUserAgent() {
        DynamoDbClient client = defaultClientBuilder().build();

        assertThatThrownBy(() -> client.listTables(r -> r
            .overrideConfiguration(o -> o.putHeader(USER_AGENT_HEADER, ""))))
            .hasMessageContaining(INTERCEPTOR_STOP_MESSAGE);

        assertUserAgentContains("");
    }

    @Test
    void executeRequest_withRequestLevelEmptyListUserAgent_shouldResultInNoUserAgent() {
        DynamoDbClient client = defaultClientBuilder().build();

        assertThatThrownBy(() -> client.listTables(r -> r
            .overrideConfiguration(o -> o.putHeader(USER_AGENT_HEADER, Collections.emptyList()))))
            .hasMessageContaining(INTERCEPTOR_STOP_MESSAGE);

        List<String> userAgentList = getCapturedUserAgentList();
        assertThat(userAgentList).isNull();
    }

    // ========== Helper Methods ==========

    private void assertUserAgentContains(String expected) {
        assertThat(getCapturedUserAgent()).contains(expected);
    }

    private void executeRequestExpectingInterception(DynamoDbClient client) {
        assertThatThrownBy(() -> client.listTables())
            .hasMessageContaining(INTERCEPTOR_STOP_MESSAGE);
    }

    private void executeRequestWithApiName(DynamoDbClient client) {
        assertThatThrownBy(() -> client.listTables(r -> r
            .overrideConfiguration(o -> o.addApiName(api -> api
                .name(TEST_API_NAME)
                .version(TEST_API_VERSION)))))
            .hasMessageContaining(INTERCEPTOR_STOP_MESSAGE);
    }

    private String getCapturedUserAgent() {
        Map<String, List<String>> headers = interceptor.context.httpRequest().headers();
        assertThat(headers).containsKey(USER_AGENT_HEADER);
        return headers.get(USER_AGENT_HEADER).get(0);
    }

    private List<String> getCapturedUserAgentList() {
        Map<String, List<String>> headers = interceptor.context.httpRequest().headers();
        return headers.get(USER_AGENT_HEADER);
    }

    private DynamoDbClientBuilder defaultClientBuilder() {
        return DynamoDbClient.builder()
                             .region(Region.US_WEST_2)
                             .credentialsProvider(StaticCredentialsProvider.create(
                                 AwsBasicCredentials.create("akid", "skid")))
                             .overrideConfiguration(c -> c.addExecutionInterceptor(interceptor));
    }

    private DynamoDbClient clientWithCustomUserAgent(String customUserAgent) {
        return DynamoDbClient.builder()
                             .region(Region.US_WEST_2)
                             .credentialsProvider(StaticCredentialsProvider.create(
                                 AwsBasicCredentials.create("akid", "skid")))
                             .overrideConfiguration(c -> c
                                 .addExecutionInterceptor(interceptor)
                                 .putHeader(USER_AGENT_HEADER, customUserAgent))
                             .build();
    }

    private DynamoDbClient clientWithCustomUserAgentList(List<String> customUserAgentList) {
        return DynamoDbClient.builder()
                             .region(Region.US_WEST_2)
                             .credentialsProvider(StaticCredentialsProvider.create(
                                 AwsBasicCredentials.create("akid", "skid")))
                             .overrideConfiguration(c -> c
                                 .addExecutionInterceptor(interceptor)
                                 .putHeader(USER_AGENT_HEADER, customUserAgentList))
                             .build();
    }

    private static class CapturingInterceptor implements ExecutionInterceptor {
        private Context.BeforeTransmission context;

        @Override
        public void beforeTransmission(Context.BeforeTransmission context, ExecutionAttributes executionAttributes) {
            this.context = context;
            throw new RuntimeException(INTERCEPTOR_STOP_MESSAGE);
        }
    }
}

```
